### PR TITLE
Improve integration tests for derive_mirror_rnda.py

### PR DESCRIPTION
**Pull request by GernotMaier**
## Overview

Improves the code related to the application `derive_mirror_rnda.py`:

- introduction of `ray_tracing` subdirectory in simtools for all ray-tracing related module (move `simtools.ray_tracing` and `simtools.psf_analysis` into this directory)
- refactor `derive_mirror_rnda.py`. Move code related to calculation of PSF of a single mirror panel from `derive_mirror_rnda.py` application to new module `ray_tracing.mirror_panel_psf`
- add functionality to `testing.compare_output` to allow for more detailed comparison of test output with reference files (especially astropy tables). Especially for this case added the possibility to compare selected columns only and apply cuts for the comparison. e.g,

```
  INTEGRATION_TESTS:
    - REFERENCE_OUTPUT_FILE: |
        tests/resources/derive_mirror_rnda_psf_mean.ecsv
      TOLERANCE: 3.e-1
      TEST_COLUMNS:
        - TEST_COLUMN_NAME: mirror_reflection_random_angle_sigma1
          CUT_COLUMN_NAME: best_fit
          CUT_CONDITION: "==True"
```

compares the output with `REFERENCE_OUTPUT_FILE`, but restrict this test to the column `mirror_reflection_random_angle_sigma1` with a mask `best_fit==True`. Required to add the corresponding keyword `TEST_COLUMNS` to the metaschema `simtools/schemas/integration_tests_config.metaschema.yml`. Added the required reference test files for all integration tests for `derive_mirror_rnda`.

Improved unit tests for ray tracing classes (consistent mocking to avoid DB access; unit tests for new module `ray_tracing.mirror_panel_psf`

A future PR will add the functionality to write the `mirror_reflection_random_angle` as result from the application (after merging PR #1215).

## Fixes

- `derive_mirror_rnda_psf_mean.yml` used MSTS/MSTN model and LST mirror list. Changed to LSTS/LSTN model

## Copilot summary

This pull request includes significant updates to the documentation and codebase related to PSF analysis, ray tracing, and mirror panel simulations. The most important changes include the removal of the `psf_analysis` module, updates to the `ray_tracing` documentation, and refactoring of the `derive_mirror_rnda.py` script for improved clarity and functionality.

### Documentation Updates:
* Removed the `psf_analysis` module documentation from `docs/source/api-reference/psf_analysis.md`.
* Updated `ray_tracing` documentation to include `psf_analysis` and `mirror_panel_psf` modules in `docs/source/api-reference/ray_tracing.md`.

### Code Refactoring:
* Refactored `derive_mirror_rnda.py` to use the `MirrorPanelPSF` class, simplifying the main function and removing redundant methods [[1]](diffhunk://#diff-f7d9299b453d898d5f4c334cb9c7c20b6659365cba26830b61daf60a55db289fL136-R139) [[2]](diffhunk://#diff-f7d9299b453d898d5f4c334cb9c7c20b6659365cba26830b61daf60a55db289fL333-R222).
* Updated imports in several scripts to reflect the new module structure (`simtools/applications/derive_psf_parameters.py`, `simtools/applications/validate_cumulative_psf.py`, `simtools/applications/validate_optics.py`) [[1]](diffhunk://#diff-6d8fe5981c82acb5871276e4931cc4b8045311f25abfd04ba8eda3b5cac177c1L96-R96) [[2]](diffhunk://#diff-86f09d75a2287c99d126dcd9eba3953a8afeadfe0700cd5611f7d746f9433e32L84-R84) [[3]](diffhunk://#diff-0863b40a69d7c64c0edd4ec7b75d2df2752190fbcade4896dedbe52c24d57f6eL80-R80).

### Bug Fixes:
* Fixed a bug in `telescope_model.py` related to the handling of mirror list file paths [[1]](diffhunk://#diff-8c441199ccfd7545fe8c5ea527c07b8ebb9f2e5dded663c1725aa92e040a6fd2L97-R97) [[2]](diffhunk://#diff-8c441199ccfd7545fe8c5ea527c07b8ebb9f2e5dded663c1725aa92e040a6fd2R132-R143).

These changes collectively improve the maintainability and clarity of the codebase and documentation.

### CODE REVIEW

File: simtools/applications/derive_mirror_rnda.py [hunk 1]
```diff
@@ -82,12 +83,12 @@
...
+    use_random_focal_length (activation mode, optional)
+        Use random focal lengths, instead of the measured ones. The argument random_focal_length
+        can be used to replace the default random_focal_length from the model.
+    random_focal_length (float, optional)
         Value of the random focal lengths to replace the default random_focal_length. Only used if
-         use_random_flen is activated.
+        'use_random_focal_length' is activated.
     no_tuning (activation mode, optional)
         Turn off the tuning - A single case will be simulated and plotted.
     test (activation mode, optional)
```
Comment (tobiaskleiner) [inline simtools/applications/derive_mirror_rnda.py:94]: Does the example in the docstring still runs for you?

I am getting: `ValueError: Invalid model version None in DB CTAO-Simulation-Model-v0-4-0 (allowed are ['6.0.0', '5.0.0'])
`
and if specifying additionally the model version:
`FileNotFoundError: [Errno 2] No such file or directory: 'MLTdata-preproduction.ecsv'`

Truncated output:
`INFO::db_handler(l145)::_find_latest_simulation_model_db::Updated the DB simulation model to the latest version CTAO-Simulation-Model-v0-4-0
Traceback (most recent call last):
  File "/workdir/env/bin/simtools-derive-mirror-rnda", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/workdir/external/simtools/simtools/applications/derive_mirror_rnda.py", line 219, in main
    panel_psf = MirrorPanelPSF(label, args_dict, db_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workdir/external/simtools/simtools/ray_tracing/mirror_panel_psf.py", line 47, in __init__
    self._get_psf_containment()
  File "/workdir/external/simtools/simtools/ray_tracing/mirror_panel_psf.py", line 104, in _get_psf_containment
    _psf_list = Table.read(
                ^^^^^^^^^^^
 `

Comment (GernotMaier) [reply simtools/applications/derive_mirror_rnda.py:94]: Fixed the docstring example.

As a side note: I don't like the docstrings examples at all. I see the value for the documentation, but at this point there are not tested automatically and they are a duplication of the examples given in the integration test. Ideal would be to sync them somehow.

Comment (tobiaskleiner) [reply simtools/applications/derive_mirror_rnda.py:94]: Thanks it runs now (but seing a Warning and the filename looks odd)! I wonder how it generated the filename in such a way. I would expect it should be None from looking at the ModelDataWriter if no output filename is supplied?
Yes agree it would be ideal if it generates the docstring examples from the configuration files.


```
Measured D80:
Mean = 1.400 cm

Simulated D80:
Mean = 1.400 cm, StdDev = 0.017 cm

mirror_random_reflection_angle
Previous value = 0.003000
New value = 0.003905

WARNING::metadata_collector(l120)::get_data_model_schema_file_name::No schema file found.
INFO::model_data_writer(l143)::write::Writing data to /workdir/external/simtools/simtools-output/d-2024-11-01/371423d8-5f1b-4c64-998a-2439c3c043a4-derive_mirror_rnda.ecsv
```

Comment (GernotMaier) [reply simtools/applications/derive_mirror_rnda.py:94]: You mean "371423d8-5f1b-4c64-998a-2439c3c043a4-derive_mirror_rnda.ecsv
"? This is the default when no metadata is given: a newly generated UUID plus the tool name (something which is in simtools since a long time). Important to note is that this is intermediate output not used for anything. We need to implement the writing of the model parameter (simtools-style), which requires first merging PR #1215.  

Also ignore the warning from the metadata collector (should have mentioned this above) - this will be fixed in the next step (this PR is about testing).

File: simtools/ray_tracing/ray_tracing.py [hunk 2]
```diff
@@ -108,38 +101,39 @@ def __repr__(self):
 
     def _initialize_config(
```
Comment (tobiaskleiner) [inline simtools/ray_tracing/ray_tracing.py:102]: A little bit hard to follow the code here: the config is not initialized here but actually now later directly in SimulatorRayTracing. Maybe renaming/deleting the function and having functions that each set self.mirror_number and self.source_distance ?

File: simtools/ray_tracing/ray_tracing.py [hunk 3]
```diff
@@ -459,29 +443,12 @@ def _analyze_image(
...
         return (
             this_off_axis * u.deg,
-            containment_diameter_cm * u.cm,
-            containment_diameter_deg * u.deg,
-            eff_area * u.m * u.m,
-            eff_flen * u.cm,
+            image.get_psf(containment_fraction, "cm") * u.cm,
+            image.get_psf(containment_fraction, "deg") * u.deg,
+            image.get_effective_area(tel_transmission) * u.m * u.m,
+            np.nan if this_off_axis == 0 else image.centroid_x / tan(this_off_axis * pi / 180.0),
```
Comment (tobiaskleiner) [inline simtools/ray_tracing/ray_tracing.py:451]: We should also use centroid_y at one point for the effective focal length: See #981

**Review by tobiaskleiner — COMMENTED** [review]
Thanks for the PR @GernotMaier, I have added some comments.

File: simtools/ray_tracing/ray_tracing.py (hunk 3 above)
Comment (GernotMaier) [reply simtools/ray_tracing/ray_tracing.py:451]: Yes, agree. But not as part of this PR.

File: simtools/ray_tracing/ray_tracing.py [hunk 4]
```diff
@@ -691,10 +637,10 @@ def _generate_file_name(
             suffix=suffix,
             site=self.telescope_model.site,
             telescope_model_name=self.telescope_model.name,
-            source_distance=self.config.source_distance,
-            zenith_angle=self.config.zenith_angle,
+            source_distance=None if self.single_mirror_mode else self.mirrors[0]["source_distance"],
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply simtools/ray_tracing/ray_tracing.py:640]: The reason why this is connect is that for single mirror mode,  each mirror panel is characterized by a measured focal length and we want to use this to determine the source distances (2x focal length). This means that source distances and mirror characteristics. Does this explain the implementation choice?

File: simtools/ray_tracing/ray_tracing.py (hunk 2 above)
Comment (GernotMaier) [reply simtools/ray_tracing/ray_tracing.py:102]: Renamed this function to `_initialize_mirror_configuration` to make it clearer. Note the comment on source_distance below.

**Comment by GernotMaier** [comment]
@tobiaskleiner - addressed your comments and merged with main. Let me know if there is anything else.

File: tests/unit_tests/ray_tracing/test_mirror_panel_psf.py [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    mirror_panel_psf.print_results()
+    out = capsys.readouterr().out
+    assert "StdDev" in out
+    assert "New value" in out
+
+    mirror_panel_psf.args_dict["psf_measurement_containment_sigma"] = None
+    mirror_panel_psf.print_results()
+    out = capsys.readouterr().out
+    assert "StdDev = 0.010 cm" not in out
+    assert "New value" in out
```
Comment (tobiaskleiner) [inline tests/unit_tests/ray_tracing/test_mirror_panel_psf.py:243]: Not fully sure about this one here. We use the same out as above without deleting in between and check for the same string. Probably need to be cleared in between?

File: tests/unit_tests/ray_tracing/test_mirror_panel_psf.py [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+import logging
+from unittest.mock import patch
+
+import astropy.units as u
+import pytest
+
+from simtools.ray_tracing.mirror_panel_psf import MirrorPanelPSF
+
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
```
[the remark this answers is not in the document]
Comment (GernotMaier) [reply tests/unit_tests/ray_tracing/test_mirror_panel_psf.py]: Removed.

**Review by tobiaskleiner — APPROVED** [review]
@GernotMaier see minor comment, besides all good - thanks!

File: tests/unit_tests/ray_tracing/test_mirror_panel_psf.py (hunk 5 above)
Comment (GernotMaier) [reply tests/unit_tests/ray_tracing/test_mirror_panel_psf.py:243]: Surprisingly little in the documentation for capsys, but found this in the source code:

```
def readouterr(self) -> CaptureResult[AnyStr]:
        """Read and return the captured output so far, resetting the internal
        buffer.
```

So the buffers are clearer.

I also checked this with simple print statements for above test function and confirm that buffers are cleared.

Comment (tobiaskleiner) [reply tests/unit_tests/ray_tracing/test_mirror_panel_psf.py:243]: Okey good, thanks for checking!

Comment (tobiaskleiner) [reply tests/unit_tests/ray_tracing/test_mirror_panel_psf.py:243]: And you are free to merge
